### PR TITLE
Add basic support for parameters in general participating in rename

### PR DESCRIPTION
**Pull request by mlorbetske**
Value transforms are not supported at this point.

Fixes an issue where the end to end test harness was being recompiled on every call (slowing things down substantially)

Adds the list of created files to the output for failed end to end tests

Switches to async reading of the output streams from the end to end test harness

Standardizes all path separators in the physical file system to be forward slashes (to deal with the parent finding logic)

Extends the template parameter interface (with a non-cache intended property) to accommodate the path portion to rename

Moves the calling for the end to end test harness to a base class

### CODE REVIEW

**Comment by mlorbetske** [comment]
This adds the basics needed for #410 and #396 but will still need #141 to be finished to be a fully working solution

File: src/Microsoft.TemplateEngine.Orchestrator.RunnableProjects/SimpleConfigModel.cs [hunk 1]
```diff
@@ -896,6 +866,59 @@ private List<FileSourceMatchInfo> EvaluateSources(IParameterSet parameters, IVar
...
+                        fileRenameMappings.Add(new KeyValuePair<string, string>(p.FileRename, s));
+                    }
+                }
+            }
+
+            foreach (KeyValuePair<string, string> entry in coreRenames)
+            {
+                foreach (KeyValuePair<string, string> rename in fileRenameMappings)
+                {
+                    string outRel = entry.Value.Replace(rename.Key, rename.Value);
```
Comment (seancpeters) [inline src/Microsoft.TemplateEngine.Orchestrator.RunnableProjects/SimpleConfigModel.cs]: Nit: I'm not sure what this variable name "outRel" is supposed to mean.

Comment (mlorbetske) [inline src/Microsoft.TemplateEngine.Orchestrator.RunnableProjects/SimpleConfigModel.cs]: Yep, these are the "output relative path" and "template relative path" values for a particular piece of content. I'll rename them

File: test/Microsoft.TemplateEngine.EndToEndTestHarness/Program.cs [hunk 2]
```diff
@@ -198,8 +205,9 @@ private static void FirstRun(IEngineEnvironmentSettings environmentSettings, IIn
 
             string packages = Path.Combine(dir, "..", "..", "..", "..", "..", "artifacts", "packages") + Path.DirectorySeparatorChar + "*";
             string templates = Path.Combine(dir, "..", "..", "..", "..", "..", "template_feed") + Path.DirectorySeparatorChar;
+            string testTemplates = Path.Combine(dir, "test_templates") + Path.DirectorySeparatorChar;
             installer.InstallPackages(new[] { packages });
-            installer.InstallPackages(new[] { templates });
+            installer.InstallPackages(new[] { templates, testTemplates });
```
Comment (seancpeters) [inline test/Microsoft.TemplateEngine.EndToEndTestHarness/Program.cs]: Should we maybe not install the real templates for tests? Changes to them could conceivably cause side-effects in the tests.

**Pull request merged**